### PR TITLE
[0.18] Refactor StreamApp and fix ServerBuilder.serve

### PR DESCRIPTION
**Pull request by aeons**
I changed `ServerBuilder.serve` back to something that makes sense.

There's also yet another attempt at refactoring `StreamApp`. It tests and I have checked it in a small example app as well.

### CODE REVIEW

File: core/src/main/scala/org/http4s/util/StreamApp.scala [hunk 1]
```diff
@@ -14,34 +14,35 @@ abstract class StreamApp[F[_]](implicit F: Effect[F]) {
...
-      }))
-      res <- stream(args.toList, shutdownRequested.set(true))
-        .interruptWhen(shutdownRequested)
-        .onFinalize(halted.set(true))
-        .as(())
-    } yield res
-    s.drain
-  }
+  /** Adds a shutdown hook that interrupts the stream and waits for it to finish */
+  private[util] def addShutdownHook(requestShutdown: Signal[F, Boolean], halted: Signal[IO, Boolean]): F[Unit] =
```
Comment (rossabaker) [inline core/src/main/scala/org/http4s/util/StreamApp.scala]: Who calls this outside of `StreamApp`?

File: core/src/main/scala/org/http4s/util/StreamApp.scala [hunk 2]
```diff
@@ -14,34 +14,35 @@ abstract class StreamApp[F[_]](implicit F: Effect[F]) {
...
+    F.delay(sys.addShutdownHook {
+      val hook = requestShutdown.set(true).runAsync(_ => IO.unit) >> halted.discrete.takeWhile(_ == false).run
+      hook.unsafeRunSync()
```
Comment (rossabaker) [inline core/src/main/scala/org/http4s/util/StreamApp.scala]: Why are we doing this?  To block the shutdown hook until halt is confirmed?

File: server/src/main/scala/org/http4s/server/ServerBuilder.scala [hunk 3]
```diff
@@ -50,7 +50,7 @@ trait ServerBuilder[F[_]] {
    * that runs for the rest of the JVM's life.
    */
   final def serve(implicit F: Async[F]): Stream[F, Nothing] =
-    Stream.bracket(start)(_ => Stream.empty, _.shutdown)
+    Stream.bracket(start)((_: Server[F]) => Stream.eval_(F.async[Unit](_ => ())), _.shutdown)
```
Comment (aeons) [inline server/src/main/scala/org/http4s/server/ServerBuilder.scala]: ssshhh

File: core/src/main/scala/org/http4s/util/StreamApp.scala (hunk 1 above)
Comment (aeons) [inline core/src/main/scala/org/http4s/util/StreamApp.scala]: Noone, was just a sloppy copy/paste from the `doMain`.

File: core/src/main/scala/org/http4s/util/StreamApp.scala (hunk 2 above)
Comment (aeons) [inline core/src/main/scala/org/http4s/util/StreamApp.scala]: The whole thing? It was there before I ported it to cats.

I assume the idea is to allow the server stream to finish before the JVM dies, when a shutdown is started from somewhere outside.

File: server/src/main/scala/org/http4s/server/ServerBuilder.scala (hunk 3 above)
Comment (rossabaker) [inline server/src/main/scala/org/http4s/server/ServerBuilder.scala]: Once it's on a release branch, it's everybody's fault. :)

**Pull request merged**